### DISCUSSION TITLE
Rework permissions for PTR zones

### DISCUSSION THREAD

**Discussion by jhg03a** [vinyldns/vinyldns#1106] [Ideas]
PTR zones are a beast of a different color from other DNS zones.  While it makes sense for groups to create their own subdomains and thus have easy access to their own records without admin rights, that model doesn't work for PTR zones.  Bodies allocating IP addresses aren't necessarily going to allocate full blocks to VinylDNS groups at a time.  Thus your only option is to use the shared zone models.  This creates a need for groups to form their own secondary vinylgroups of broader membership just to manage permissions to the PTR zones apart from their other zones.

My suggestion in this case is to rethink how PTR zone permissions work altogether.  Any VinylDNS user with write permissions in a specific zone should be able to ubiquitously create PTR records without PTR zone membership.  Instead it should permit only ptrdnames for zones they do have permissions to with an associated vinyldns group owner.  In the event where DNS records weren't properly cleaned up between IP reallocations, being able to still insert current ptrdnames is useful.  Additionally, I would want a better way to locate the vinyldns group to which the old ptrdrecord belongs for a request to be cleaned up.

**Comment by remerle** [comment]
This is more of an administrative request for operators of VinylDNS instances than a technical limitation. Users of VinylDNS should be permitted to assign ownership to reverse zones as their business needs dictate.

We cannot safely assume that the owner of an unrelated zone is permitted to "claim" the reverse zone record. We also cannot assume that VinylDNS is the source of truth, so we cannot determine if there are any other forward records that would correspond to the reverse record.

To avoid unsafe assumptions, I propose a third type of zone; an "open" zone. We have "private" and "shared" at present, so "open" would simply be a "shared" zone but without ownership. This would allow VinylDNS operators to opt-in to the behavior of allowing anyone to change PTR records _(or records in any zone they wish)_ and rely on the audit logs to keep everyone honest.

**Comment by remerle** [comment]
In general, and in my opinion, the entire ownership structure needs to be revisited. There are too many opportunities for ownership to stagnate over time. Also, I believe that the "email the owner to request access" model is less than ideal and could use to have a more formal process within VinylDNS.

A couple of ideas worth considering:

1. A record in a shared zone can be owned, but ownership is time-limited. 
	- Owners could renew their interest in records as they come to expire
  	- Failing to renew their ownership causes the record to return to an "unowned" state
  	- This would all be automated
1. An in-built mechanism for requesting modifications from owners.
	- It should not be an onerous ordeal to change records in shared zones if they are owned. 
  		- We shouldn't expect users to contact all of the owners and track their responses
  	- VinylDNS should provide a request queue for owners and requestors alike
  		- Owners will be notified if there's a request to change one of their shared records. Likewise, the requestor will have a centralized view of all of the outstanding requests
  		- Failure of an owner to respond to a request (and reminders) after a configurable period (e.g., 72 hours) should result in the request being auto-permitted.
  		- If a change is auto-permitted, the full change details should be sent to the owner via notification
  		- Each request to change a shared record will require approval, and the requested change should be immutable. If the request requires a change, the user should have to withdraw the original request and create a new one. This is to prevent a change from being approved but then being altered right before its application.
1. An in-built mechanism for ownership transfer
    - This is the same as 2) above, but not for permitting a change, instead for transferring ownership to a new group
    - This should not be auto-approved, but an administrator could intervene and approve the transfer
    - This could apply to zones as well as records
    - This could apply to a bulk set of records (maybe even based on some regex matching the FQDN?)

In general, providing users with the ability to self-service ownership requests would go a long way to reducing the support burden on the operators.

**Comment by jhg03a** [comment]
Regardless of if you go with the first or second option for PTR zones, automation users need to have a mechanism to resolve the zone to its zoneid so they can continue to make individual change requests and query recordset status.  Going the second option route is going to be problematic from an automation standpoint because it means there's more than a success/fail return state for a single record change.  Submitting batch changes blindly doesn't support idempotentcy goals.

**Reply by remerle** [reply]
I would like to move to a model where zone data can be read by anyone. We effectively have this with the global recordset search anyway.

Modification of PTR records still remains an open issue that will need some consideration.

@jhg03a Are there any interim steps that you think would be helpful for your automation tasks? For instance, making zone/record data more readily viewable.

**Reply by jhg03a** [reply]
When attempting to audit my DNS records by evaluating by my IP spaces, having increased visibility into other zones would make a significant difference.  For now, I can just examine what are in zones I can see, then perform reverse dns lookups to guess at things outside my vinyldns zones that could exist and that I should clean up.

**Comment by jhg03a** [comment]
I generally try to use the Terraform plugin when possible since it handles state and idempotency fairly well.  For specific tasks I typically use either the golang library or raw API calls in python/ansible.  My current scope has me unfortunately in both worlds somewhat.  I can use Terraform to manage records for zones in which I am an owner, but not in shared zones (i.e. high value zones) where I don't have high enough permissions.  So just "can I see it when I ask for my groups" isn't enough.

So I have to split the records I'm attempting to enforce into separate groups, one where it works with Terraform and is pretty easy.  The latter being a set of records I have to effectively reimplement a minimalist version of Terraform myself and make use of some other means to submit batch changes instead (since those don't rely on the zone ownership model in the same way).  PTR recordsets are where the latter mostly comes into play, however it's also relevant in high value zones as well.

It's a far from simple and error prone solution to manage the state myself and use a batch api that introduces new ways for individual records to fail or be caught in an approval state of some kind.  It also will create a fair amount of clutter with regard to my batch change list over time.  It's unfortunately the best idea I could come up with at the time other than just abandoning the idea of enforcing records outside my zone ownerships.  It's a very different way of thinking to say "do what is neccessary if anything to make these thousands of records look like this" every few days/hours/weeks/etc as opposed to "I have this set of known things that need to change one time to be successful".

**Comment by jhg03a** [comment]
With how the system is today, I put in a quality of life UI enhancement #1122.  If we do look at making PTR zones into a beast of a different color with permissions, I'm not sure it makes sense to display.  With thousands of PTR zones I'm guessing I might have, including those in the zones list of the UI would be a lot of mess to sort through.  I'm guessing this is already the case for admin level users today as well.  Programmatically via the API it's fine to list them provided that the client libraries that have built in pagination expose retry logic or the underlying functions they invoke (https://github.com/vinyldns/go-vinyldns/issues/88).
